<a href="https://colab.research.google.com/github/Harsh0487/Harsh_INFO5731_fall2024/blob/main/INFO5731_Assignment_1_Spring_2024.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **INFO5731 Assignment One**

This assignment includes range of programming tasks and data analysis exercises, aiming to enhance your Python skills and proficiency in string manipulation, numerical computations, and data science. Follow the instructions carefully and design well-structured Python programs to address each question.

Late Submission will have a penalty of 10% reduction for each day after the deadline.

# Question 1 (10 points).

The formula that describes the volume V of a hemisphere with radius r is the following:

$ V=\frac{2}{3}\ast\ \pi\ast\ r^3 $

Write a program to calculate the value of V when r is in the range of even numbers from 1-20. Output the result in the following format:

	r	V
	2	…
	4	…
	6	…
	.
	.
	.
	20	…


In [1]:
import math


def calculate_hemisphere_volume(radius):
    return (2/3) * math.pi * (radius ** 3)


even_radii = range(2, 21, 2)


for radius in even_radii:
    volume = calculate_hemisphere_volume(radius)
    print(f"Radius: {radius}, Volume: {volume:.2f}")



Radius: 2, Volume: 16.76
Radius: 4, Volume: 134.04
Radius: 6, Volume: 452.39
Radius: 8, Volume: 1072.33
Radius: 10, Volume: 2094.40
Radius: 12, Volume: 3619.11
Radius: 14, Volume: 5747.02
Radius: 16, Volume: 8578.64
Radius: 18, Volume: 12214.51
Radius: 20, Volume: 16755.16


# Question 2 (20 points)

(a) Write a Python program to generate a sequence of 12 random numbers falling within the range of 50 to 150.

(b) Arrange the generated numbers in ascending order. Display the sorted list.

(c) Extend the program to arrange the numbers in descending order as well. Present the sorted list accordingly.

(d) Compute and output the average of the entire group of generated numbers.

(e) Enhance the program to calculate and display the average of the first 6 numbers from both the ascending and descending sorted lists.

In [10]:

import random
min_value = 50
max_value = 150
num_random_numbers = 12
random_numbers = [random.randint(min_value, max_value) for _ in range(num_random_numbers)]
print("Random numbers:")
for i, num in enumerate(random_numbers, start=1):
    print(f"Number {i}: {num}")

    #Q2-#(b)

import random
min_value = 50
max_value = 150


num_random_numbers = 12
random_numbers = [random.randint(min_value, max_value) for _ in range(num_random_numbers)]

print("Unsorted random numbers:")
for i, num in enumerate(random_numbers, start=1):
    print(f"Number {i}: {num}")

sorted_numbers = sorted(random_numbers)


print("\nSorted random numbers:")
for i, num in enumerate(sorted_numbers, start=1):
    print(f"Number {i}: {num}")



Random numbers:
Number 1: 79
Number 2: 72
Number 3: 74
Number 4: 54
Number 5: 80
Number 6: 70
Number 7: 77
Number 8: 146
Number 9: 67
Number 10: 118
Number 11: 92
Number 12: 102
Unsorted random numbers:
Number 1: 62
Number 2: 87
Number 3: 86
Number 4: 104
Number 5: 125
Number 6: 127
Number 7: 107
Number 8: 143
Number 9: 145
Number 10: 146
Number 11: 93
Number 12: 111

Sorted random numbers:
Number 1: 62
Number 2: 86
Number 3: 87
Number 4: 93
Number 5: 104
Number 6: 107
Number 7: 111
Number 8: 125
Number 9: 127
Number 10: 143
Number 11: 145
Number 12: 146


In [11]:
#Q2-#(c)

import random
min_value = 50
max_value = 150
num_random_numbers = 12


random_numbers = [random.randint(min_value, max_value) for _ in range(num_random_numbers)]


print("Unsorted random numbers:")
for i, num in enumerate(random_numbers, start=1):
    print(f"Number {i}: {num}")

sorted_numbers_ascending = sorted(random_numbers)


sorted_numbers_descending = sorted(random_numbers, reverse=True)

print("\nSorted random numbers (Ascending):")
for i, num in enumerate(sorted_numbers_ascending, start=1):
    print(f"Number {i}: {num}")

print("\nSorted random numbers (Descending):")
for i, num in enumerate(sorted_numbers_descending, start=1):
    print(f"Number {i}: {num}")



Unsorted random numbers:
Number 1: 133
Number 2: 139
Number 3: 141
Number 4: 138
Number 5: 139
Number 6: 146
Number 7: 112
Number 8: 52
Number 9: 87
Number 10: 61
Number 11: 79
Number 12: 129

Sorted random numbers (Ascending):
Number 1: 52
Number 2: 61
Number 3: 79
Number 4: 87
Number 5: 112
Number 6: 129
Number 7: 133
Number 8: 138
Number 9: 139
Number 10: 139
Number 11: 141
Number 12: 146

Sorted random numbers (Descending):
Number 1: 146
Number 2: 141
Number 3: 139
Number 4: 139
Number 5: 138
Number 6: 133
Number 7: 129
Number 8: 112
Number 9: 87
Number 10: 79
Number 11: 61
Number 12: 52


In [12]:
#Q2-#(d)

import random


min_value = 50
max_value = 150


num_random_numbers = 12


random_numbers = [random.randint(min_value, max_value) for _ in range(num_random_numbers)]


print("Unsorted random numbers:")
for i, num in enumerate(random_numbers, start=1):
    print(f"Number {i}: {num}")


sorted_numbers_ascending = sorted(random_numbers)


sorted_numbers_descending = sorted(random_numbers, reverse=True)


print("\nSorted random numbers (Ascending):")
for i, num in enumerate(sorted_numbers_ascending, start=1):
    print(f"Number {i}: {num}")

print("\nSorted random numbers (Descending):")
for i, num in enumerate(sorted_numbers_descending, start=1):
    print(f"Number {i}: {num}")


average = sum(random_numbers) / num_random_numbers
print(f"\nAverage of the random numbers: {average:.2f}")

#Q2-#(e).

import random

min_value = 50
max_value = 150


num_random_numbers = 12


random_numbers = [random.randint(min_value, max_value) for _ in range(num_random_numbers)]

print("Unsorted random numbers:")
for i, num in enumerate(random_numbers, start=1):
    print(f"Number {i}: {num}")
sorted_numbers_ascending = sorted(random_numbers)
sorted_numbers_descending = sorted(random_numbers, reverse=True)
print("\nSorted random numbers (Ascending):")
for i, num in enumerate(sorted_numbers_ascending, start=1):
    print(f"Number {i}: {num}")

print("\nSorted random numbers (Descending):")
for i, num in enumerate(sorted_numbers_descending, start=1):
    print(f"Number {i}: {num}")

average = sum(random_numbers) / num_random_numbers

print(f"\nAverage of the random numbers: {average:.2f}")
average_ascending_first_6 = sum(sorted_numbers_ascending[:6]) / 6


print(f"\nAverage of the first 6 numbers from the ascending sorted list: {average_ascending_first_6:.2f}")


average_descending_first_6 = sum(sorted_numbers_descending[:6]) / 6
print(f"\nAverage of the first 6 numbers from the descending sorted list: {average_descending_first_6:.2f}")


Unsorted random numbers:
Number 1: 114
Number 2: 98
Number 3: 131
Number 4: 82
Number 5: 106
Number 6: 80
Number 7: 73
Number 8: 95
Number 9: 72
Number 10: 73
Number 11: 128
Number 12: 103

Sorted random numbers (Ascending):
Number 1: 72
Number 2: 73
Number 3: 73
Number 4: 80
Number 5: 82
Number 6: 95
Number 7: 98
Number 8: 103
Number 9: 106
Number 10: 114
Number 11: 128
Number 12: 131

Sorted random numbers (Descending):
Number 1: 131
Number 2: 128
Number 3: 114
Number 4: 106
Number 5: 103
Number 6: 98
Number 7: 95
Number 8: 82
Number 9: 80
Number 10: 73
Number 11: 73
Number 12: 72

Average of the random numbers: 96.25
Unsorted random numbers:
Number 1: 121
Number 2: 111
Number 3: 92
Number 4: 67
Number 5: 93
Number 6: 127
Number 7: 124
Number 8: 148
Number 9: 119
Number 10: 54
Number 11: 106
Number 12: 112

Sorted random numbers (Ascending):
Number 1: 54
Number 2: 67
Number 3: 92
Number 4: 93
Number 5: 106
Number 6: 111
Number 7: 112
Number 8: 119
Number 9: 121
Number 10: 124
Number 

# Question 3 (20 points).

Create a Python program that encompasses various string manipulation task:

(a) Prompt the user to input two sentences.

(b) Transform each sentence into a list, where each element corresponds to a word in the sentence. For example, if the sentences are "Hello, how are you doing today?" and "This is an example sentence for the program!"

The program should output:
['Hello', ',', 'how', 'are', 'you', 'doing', 'today', '?', 'This', 'is', 'an', 'example', 'sentence', 'for', 'the', 'program', '!']

(c) Compare the two sentences. Print out a list of words that occur in both sentences.

(d) Remove any punctuation from the sentence and print the sanitized sentences. It should remove at least the following characters, but it can remove more: period(“.”), comma (“,”), semicolon (“;”), and colon (“:”).

(e) Count the number of vowels in each of the sentences. Print out each vowel and the occurrences  in the each of the two sentences separately, such as:

Sentence1: a:2, e:2, i:0, etc.
Sentence2: a:1, e:2, i:2, etc.

Design a well-structured Python program to execute these tasks efficiently.

In [14]:
import string

def get_sentences():
    """Prompt the user to input two sentences."""
    sentence1 = input("Please enter the first sentence: ")
    sentence2 = input("Please enter the second sentence: ")
    return sentence1, sentence2

def split_sentences_into_words(sentence1, sentence2):
    """Transform each sentence into a list, where each element corresponds to a word in the sentence."""
    words = sentence1.split() + sentence2.split()
    return words

def compare_sentences(sentence1, sentence2):
    """Compare the two sentences. Print out a list of words that occur in both sentences."""
    set1 = set(sentence1.split())
    set2 = set(sentence2.split())
    common_words = set1.intersection(set2)
    return list(common_words)

def remove_punctuation(sentence1, sentence2):
    """Remove any punctuation from the sentence and print the sanitized sentences."""
    translator = str.maketrans('', '', string.punctuation)
    sanitized_sentence1 = sentence1.translate(translator)
    sanitized_sentence2 = sentence2.translate(translator)
    return sanitized_sentence1, sanitized_sentence2

def count_vowels(sentence1, sentence2):
    """Count the number of vowels in each of the sentences."""
    vowels = 'aeiou'
    vowel_counts1 = {vowel: sentence1.lower().count(vowel) for vowel in vowels}
    vowel_counts2 = {vowel: sentence2.lower().count(vowel) for vowel in vowels}
    return vowel_counts1, vowel_counts2

def main():
    sentence1, sentence2 = get_sentences()
    print("\nYou entered the following sentences:")
    print("Sentence 1:", sentence1)
    print("Sentence 2:", sentence2)

    words = split_sentences_into_words(sentence1, sentence2)
    print("\nCombined list of words:")
    print(words)

    common_words = compare_sentences(sentence1, sentence2)
    print("\nCommon words:")
    print(common_words)

    sanitized_sentence1, sanitized_sentence2 = remove_punctuation(sentence1, sentence2)
    print("\nSanitized sentences:")
    print("Sentence 1:", sanitized_sentence1)
    print("Sentence 2:", sanitized_sentence2)

    vowel_counts1, vowel_counts2 = count_vowels(sentence1, sentence2)
    print("\nVowel counts:")
    print("Sentence 1:", ', '.join(f'{vowel}:{count}' for vowel, count in vowel_counts1.items()))
    print("Sentence 2:", ', '.join(f'{vowel}:{count}' for vowel, count in vowel_counts2.items()))

if __name__ == "__main__":
    main()




Please enter the first sentence: hello how are you?
Please enter the second sentence: what is going on buddy today?

You entered the following sentences:
Sentence 1: hello how are you?
Sentence 2: what is going on buddy today?

Combined list of words:
['hello', 'how', 'are', 'you?', 'what', 'is', 'going', 'on', 'buddy', 'today?']

Common words:
[]

Sanitized sentences:
Sentence 1: hello how are you
Sentence 2: what is going on buddy today

Vowel counts:
Sentence 1: a:1, e:2, i:0, o:3, u:1
Sentence 2: a:2, e:0, i:2, o:3, u:1


# Question 4 (20 Points)

Design and implement a simple command-line calculator program. The program should handle both integers and floating-point numbers.

(a) Write program that takes a mathematical expression as input and evaluates it. The program should support addition, subtraction, multiplication, and division.

(b) Extend the calculator to handle parentheses in mathematical expressions. Implement the necessary logic to evaluate expressions with nested parentheses.

(c) Enhance your calculator to include additional mathematical operations such as exponentiation, square root, and modulus.

(d) Implement error handling in the calculator program. Handle cases such as division by zero and invalid input expressions. Provide meaningful error messages.

(e) Implement a command to exit the calculator.

In [23]:
import numexpr as ne
import math

def calculate_expression(expression):
    try:
        result = ne.evaluate(expression)
        return result
    except ZeroDivisionError:
        print("Error: Division by zero is not allowed.")
        return None
    except Exception as e:
        print(f"Error: Invalid input expression - {str(e)}")
        return None

def extend_calculator():
    print("Extended Calculator")
    print("1. Exponentiation (x^y)")
    print("2. Square Root (sqrt(x))")
    print("3. Modulus (x % y)")
    choice = input("Enter your choice (1/2/3): ")

    if choice == "1":
        x = float(input("Enter base: "))
        y = float(input("Enter exponent: "))
        result = math.pow(x, y)
    elif choice == "2":
        x = float(input("Enter number: "))
        if x < 0:
            print("Error: Square root of negative number is not allowed.")
            return
        result = math.sqrt(x)
    elif choice == "3":
        x = float(input("Enter dividend: "))
        y = float(input("Enter divisor: "))
        if y == 0:
            print("Error: Division by zero is not allowed.")
            return
        result = x % y
    else:
        print("Error: Invalid choice.")
        return

    print("Result:", result)

def main():
    while True:
        print("\nCommand-Line Calculator")
        print("1. Evaluate Expression")
        print("2. Extended Calculator")
        print("3. Exit")
        choice = input("Enter your choice (1/2/3): ")

        if choice == "1":
            expression = input("Enter mathematical expression: ")
            result = calculate_expression(expression)
            if result is not None:
                print("Result:", result)
        elif choice == "2":
            extend_calculator()
        elif choice == "3":
            print("Exiting calculator. Goodbye!")
            break
        else:
            print("Error: Invalid choice.")

if __name__ == "__main__":
    main()



Command-Line Calculator
1. Evaluate Expression
2. Extended Calculator
3. Exit
Enter your choice (1/2/3): 1
Enter mathematical expression: -
Error: Invalid input expression - invalid syntax (<expr>, line 1)

Command-Line Calculator
1. Evaluate Expression
2. Extended Calculator
3. Exit
Enter your choice (1/2/3): 1
Enter mathematical expression: 5-5
Result: 0

Command-Line Calculator
1. Evaluate Expression
2. Extended Calculator
3. Exit
Enter your choice (1/2/3): 3
Exiting calculator. Goodbye!


# Question 5 (30 points).

In the field of Data Science, data is often formatted as a comma-delimited (CSV) file, where each line in a file corresponds to a field's value. Refer to Chapter 9 (p. 106) for more information on comma-delimited files.

To answer this question, you will need to download the file [Assignment1_denton_housing.csv](https://raw.githubusercontent.com/unt-iialab/info5731-spring2022/main/assignments/Assignment1_denton_housing.csv) included with this assignment. The file contains statistics about housing in the city of Denton from 2008 to 2014. Write a program to implement the following questions.

(a) (5 pts) Calculate the difference in the number of Occupied Housing Units from year to year and print it. The difference must be calculated for the consecutive years such as 2008-2009, 2009-2010 etc. Finally, print the values in the ascending order.

(b) (5 pts) For all the years, calculate the percentage of housing units which are vacant an occupied. Print the results in the following format:

Year Vacant Housing Units Occupied Housing Units

    2008  30%   70%                          
	2009 ----- -----
	2010 ----- -----
    2011 ----- -----
	2012 ----- -----
    2013 ----- -----
	2014 ----- -----
  
  

(c) (10 pts) Calculate and print the valued and years in which the highest number of housing units were vacant and occupied. Print the results in the following format:  

                              Year	Value
	Vacant Housing Units	  -----   -----
	Occupied Housing Units	-----   -----

(d) (10 pts) Calculate the harmonic mean of the total housing units and print it out. You can find the information about harmonic mean here: https://ncalculators.com/statistics/harmonic-mean-calculator.htm

In [53]:
!pip install pandas
import pandas as pd
from scipy.stats import hmean

# Load the data from CSV
file_name = '/content/sample_data/Assignment1_denton_housing.csv'
df = pd.read_csv(file_name)

# (a) Calculate the difference in the number of Occupied Housing Units from year to year
df_occupied = df[df['title_field'] == 'Occupied Housing Units']
df_occupied_sorted = df_occupied.sort_values(by='year')

yearly_diff = df_occupied_sorted['value'].diff().dropna()
yearly_diff_sorted = yearly_diff.sort_values()

print("\n(a) Difference in Occupied Housing Units (sorted in ascending order):")
for i in range(len(yearly_diff_sorted)):
    print(f"{df_occupied_sorted.iloc[i+1]['year']}: {yearly_diff_sorted.iloc[i]}")

# (b) Calculate the percentage of housing units that are vacant and occupied for each year
print("\n(b) Percentage of vacant and occupied housing units:")

for year in df['year'].unique():
    total_units = df[df['year'] == year]['value'].sum()
    vacant_units = df[(df['year'] == year) & (df['title_field'] == 'Vacant Housing Units')]['value'].values[0]
    occupied_units = df[(df['year'] == year) & (df['title_field'] == 'Occupied Housing Units')]['value'].values[0]

    vacant_percent = (vacant_units / total_units) * 100
    occupied_percent = (occupied_units / total_units) * 100

    print(f"{year} {vacant_percent:.2f}% {occupied_percent:.2f}%")

# (c) Find the year and value for the highest number of vacant and occupied housing units
max_vacant = df[df['title_field'] == 'Vacant Housing Units'].sort_values(by='value', ascending=False).iloc[0]
max_occupied = df[df['title_field'] == 'Occupied Housing Units'].sort_values(by='value', ascending=False).iloc[0]

print("\n(c) Year and value of highest number of housing units:")
print(f"Vacant Housing Units      {max_vacant['year']}   {max_vacant['value']}")
print(f"Occupied Housing Units    {max_occupied['year']}   {max_occupied['value']}")

# (d) Calculate the harmonic mean of the total housing units for all years
total_units_per_year = df.groupby('year')['value'].sum()
harmonic_mean_value = hmean(total_units_per_year)

print(f"\n(d) Harmonic mean of total housing units: {harmonic_mean_value:.2f}")


(a) Difference in Occupied Housing Units (sorted in ascending order):
2009: -795.0
2010: -303.0
2011: 1404.0
2012: 1565.0
2013: 2622.0
2014: 5091.0

(b) Percentage of vacant and occupied housing units:
2014 1.94% 31.84%
2013 3.00% 31.12%
2012 2.26% 31.61%
2011 3.21% 30.99%
2010 2.77% 31.29%
2009 3.10% 31.05%
2008 3.74% 30.63%

(c) Year and value of highest number of housing units:
Vacant Housing Units      2008   4483
Occupied Housing Units    2014   46295

(d) Harmonic mean of total housing units: 130257.98


# Mandatory Question

Provide your thoughts on the assignment. What did you find challenging, and what aspects did you enjoy? Your opinion on the provided time to complete the assignment. Share any suggestions for improvement. Your feedback will be valuable for enhancing the learning experience.

In [ ]:
Assignments are better for gaining knowledge and python skills.time is enough for completion.